wasm/wazeroir/interpreter.go

```diff
@@ -0,0 +1,1502 @@
+	}
+	return
+}
+
+func (it *interpreter) callHostFunc(f *interpreterFunction, args ...uint64) {
+	tp := f.hostFn.Type()
+	in := make([]reflect.Value, tp.NumIn())
+	for i := len(in) - 1; i >= 1; i-- {
+		val := reflect.New(tp.In(i)).Elem()
+		raw := it.pop()
+		kind := tp.In(i).Kind()
+		switch kind {
+		case reflect.Float64, reflect.Float32:
+			val.SetFloat(math.Float64frombits(raw))
+		case reflect.Uint32, reflect.Uint64:
+			val.SetUint(raw)
+		case reflect.Int32, reflect.Int64:
+			val.SetInt(int64(raw))
+		}
+		in[i] = val
+	}
+
+	val := reflect.New(tp.In(0)).Elem()
+	var memory *wasm.MemoryInstance
+	if len(it.frames) > 0 {
+		memory = it.frames[len(it.frames)-1].f.moduleInstance.Memory
+	}
+	val.Set(reflect.ValueOf(&wasm.HostFunctionCallContext{Memory: memory}))
+	in[0] = val
+
+	frame := &interpreterFrame{f: f}
+	it.pushFrame(frame)
+	for _, ret := range f.hostFn.Call(in) {
+		switch ret.Kind() {
+		case reflect.Float64, reflect.Float32:
+			it.push(math.Float64bits(ret.Float()))
+		case reflect.Uint32, reflect.Uint64:
+			it.push(ret.Uint())
+		case reflect.Int32, reflect.Int64:
+			it.push(uint64(ret.Int()))
+		default:
+			panic("invalid return type")
+		}
+	}
+	it.popFrame()
+}
+
+func (it *interpreter) callNativeFunc(f *interpreterFunction) {
+	frame := &interpreterFrame{f: f}
+	memoryInst := frame.f.moduleInstance.Memory
+	globals := frame.f.moduleInstance.Globals
+	tables := f.moduleInstance.Tables
+	it.pushFrame(frame)
+	bodyLen := uint64(len(frame.f.body))
+	for frame.pc < bodyLen {
+		op := frame.f.body[frame.pc]
+		switch op.kind {
+		case OperationKindUnreachable:
+			panic("unreachable")
+		case OperationKindBr:
+			{
+				frame.pc = op.us[0]
+			}
+		case OperationKindBrIf:
+			{
+				if it.pop() > 0 {
+					it.drop(op.rs[0])
+					frame.pc = op.us[0]
+				} else {
+					it.drop(op.rs[1])
+					frame.pc = op.us[1]
+				}
+			}
+		case OperationKindBrTable:
+			{
+				if v := int(it.pop()); v < len(op.us)-1 {
+					it.drop(op.rs[v+1])
+					frame.pc = op.us[v+1]
+				} else {
+					// Default branch.
+					it.drop(op.rs[0])
+					frame.pc = op.us[0]
+				}
+			}
+		case OperationKindCall:
+			{
+				if op.f.hostFn != nil {
+					it.callHostFunc(op.f, it.stack[len(it.stack)-len(op.f.signature.InputTypes):]...)
+				} else {
+					it.callNativeFunc(op.f)
+				}
+				frame.pc++
+			}
+		case OperationKindCallIndirect:
+			{
+				index := it.pop()
+				target := it.functions[tables[op.us[0]].Table[index].Function]
+				// Type check.
+				if target.typeID != op.us[1] {
+					panic("function signature mismatch on call_indirect")
+				}
+				// Call in.
+				if target.hostFn != nil {
+					it.callHostFunc(target, it.stack[len(it.stack)-len(target.signature.InputTypes):]...)
+				} else {
+					it.callNativeFunc(target)
+				}
+				frame.pc++
+			}
+		case OperationKindDrop:
+			{
+				it.drop(op.rs[0])
+				frame.pc++
+			}
+		case OperationKindSelect:
+			{
+				c := it.pop()
+				v2 := it.pop()
```

Comment:
do we check length anywhere? ex `requireLen(4, it)` sometimes in parsing it is easier to evaluate preconditions so you can pop freely later.

Reply:
as I commented https://github.com/tetratelabs/wazero/blob/microwasm/wasm/wazeroir/compiler.go#L64-L67, the stack operations are always valid at this runtime phase -- operations are already statically checked at module validation phase before compilation : https://github.com/tetratelabs/wazero/blob/microwasm/wasm/store.go#L445